authprox/apadmin/main.go

```diff
@@ -0,0 +1,222 @@
+)
+
+var cmds = cli.Commands{
+	{
+		Name:  "add",
+		Usage: "add an external identity provider",
+		Flags: []cli.Flag{
+			cli.StringFlag{
+				Name:  "roster, r",
+				Usage: "the roster of the cothority that hosts the distributed Authentication Proxy",
+			},
+			cli.StringFlag{
+				Name:  "type",
+				Usage: "the type of validator: oidc (OpenID Connect) or tq (Tequila). Default is oidc.",
+				Value: "oidc",
+			},
+			cli.StringFlag{
+				Name:  "issuer",
+				Usage: "the URL of the identity provider (for OpenID Connect, this is the Issuer URL)",
+			},
+		},
+		Action: add,
+	},
+	{
+		Name:  "show",
+		Usage: "show the enrollments of external identity providers",
+		Flags: []cli.Flag{
+			cli.StringFlag{
+				Name:  "roster, r",
+				Usage: "the roster of the cothority that hosts the distributed Authentication Proxy",
+			},
+		},
+		Action: show,
+	},
+}
+
+var cliApp = cli.NewApp()
+
+// getDataPath is a function pointer so that tests can hook and modify this.
+var getDataPath = cfgpath.GetDataPath
+
+func init() {
+	cliApp.Name = "apadmin"
+	cliApp.Usage = "Administer authentication proxies."
+	cliApp.Version = "0.1"
+	cliApp.Commands = cmds
+	cliApp.Flags = []cli.Flag{
+		cli.IntFlag{
+			Name:  "debug, d",
+			Value: 0,
+			Usage: "debug-level: 1 for terse, 5 for maximal",
+		},
+		cli.StringFlag{
+			Name:  "config, c",
+			Value: "",
+			Usage: "path to configuration-directory",
+		},
+	}
+	cliApp.Before = func(c *cli.Context) error {
+		log.SetDebugVisible(c.Int("debug"))
+		lib.ConfigPath = c.String("config")
+		if lib.ConfigPath == "" {
+			lib.ConfigPath = getDataPath(cliApp.Name)
+		}
+		return nil
+	}
+}
+
+func main() {
+	log.ErrFatal(cliApp.Run(os.Args))
+}
+
+func show(c *cli.Context) error {
+	fn := c.String("roster")
+	if fn == "" {
+		return errors.New("--roster flag is required")
+	}
+
+	in, err := os.Open(fn)
+	if err != nil {
+		return fmt.Errorf("Could not open roster %v: %v", fn, err)
+	}
+	roster, err := readRoster(in)
+	if err != nil {
+		return err
+	}
+	cl := onet.NewClient(cothority.Suite, authprox.ServiceName)
+	req := &authprox.EnrollmentsRequest{}
+	resp := &authprox.EnrollmentsResponse{}
+	err = cl.SendProtobuf(roster.List[0], req, resp)
+	if err != nil {
+		return err
+	}
+
+	for _, x := range resp.Enrollments {
+		fmt.Println(x.Type, x.Issuer, x.Public)
+	}
+	return nil
+}
+
+func add(c *cli.Context) error {
+	is := c.String("issuer")
+	if is == "" {
+		return errors.New("--issuer flag is required")
+	}
+
+	// Load the roster
+	fn := c.String("roster")
+	if fn == "" {
+		return errors.New("--roster flag is required")
+	}
+
+	in, err := os.Open(fn)
+	if err != nil {
+		return fmt.Errorf("Could not open roster %v: %v", fn, err)
+	}
+	roster, err := readRoster(in)
+	if err != nil {
+		return err
+	}
+
+	// Get all the key material ready.
+
+	// n: how many auth proxies will be holding shares
+	n := len(roster.List)
+	T := threshold(n)
+
+	pubs := make([]kyber.Point, n)
+	for i := 0; i < n; i++ {
```

Comment:
Nobody likes `for i := range pubs`? I already had this fight with Philipp... Seems I lost.

Reply:
Yes, that's better. When you do it that way, you give the compiler enough info to be able to remove the bounds checks, because it can prove that 0 <= i  < len(pubs).